src/imitation/regularization/updaters.py

```diff
@@ -0,0 +1,139 @@
+
+        Raises:
+            ValueError: If the tolerable interval is not a tuple of length 2.
+            ValueError: if the scaling factor is not in (0, 1).
+            ValueError: if the tolerable interval is negative or not a proper interval.
+        """
+        eps = np.finfo(float).eps
+        if not (eps < scaling_factor < 1 - eps):
+            raise ValueError(
+                "scaling_factor must be in (0, 1) within machine precision.",
+            )
+        if len(tolerable_interval) != 2:
+            raise ValueError("tolerable_interval must be a tuple of length 2")
+        if not (0 <= tolerable_interval[0] < tolerable_interval[1]):
+            raise ValueError(
+                "tolerable_interval must be a tuple whose first element "
+                "is at least 0 and the second element is greater than "
+                "the first",
+            )
+
+        self.scaling_factor = scaling_factor
+        self.tolerable_interval = tolerable_interval
+
+    def __call__(
+        self,
+        lambda_: float,
+        train_loss: LossType,
+        val_loss: LossType,
+    ) -> float:
+        """Scales the lambda of the regularizer by some constant factor.
+
+        Lambda is scaled up if the ratio of the validation loss to the training loss
+        is above the tolerable interval, and scaled down if the ratio is below the
+        tolerable interval. Nothing happens if the ratio is within the tolerable
+        interval.
+
+        Args:
+            lambda_: The current value of the lambda.
+            train_loss: The loss on the training set.
+            val_loss: The loss on the validation set.
+
+        Returns:
+            The new value of the lambda.
+
+        Raises:
+            ValueError: If the loss on the validation set is not a scalar.
+            ValueError: if lambda_ is zero (will result in no scaling).
+            ValueError: if lambda_ is not a float.
+        """
+        # assert that the tensors val_loss and train_loss are both scalars
+        try:
+            assert isinstance(val_loss, float) or (
+                isinstance(val_loss, th.Tensor) and val_loss.dim() == 0
+            )
+            assert isinstance(train_loss, float) or (
+                isinstance(train_loss, th.Tensor) and train_loss.dim() == 0
+            )
+        except AssertionError as exc:
+            raise ValueError("val_loss and train_loss must be scalars") from exc
+        if np.finfo(float).eps > lambda_:
+            raise ValueError(
+                "lambda_ must not be zero. Make sure that you're not "
+                "scaling the value of lambda down too quickly or passing an "
+                "initial value of zero to the lambda parameter.",
+            )
+        if not isinstance(lambda_, float):
+            raise ValueError("lambda_ must be a float")
+
+        eps = np.finfo(float).eps
+        if train_loss < eps and val_loss < eps:
```

Comment:
This assumes `train_loss` and `val_loss`. Which it should be, but probably worth explicitly checking for, given that for most ML algorithms you can shift loss up/down by a constant value and not change things (whereas this algorithm treats zero as special because it takes the ratio of losses).

Reply:
I think you meant it assumes they are "positive" (you probably missed the word). yes, sure, will do.